File: Ghostpad-app/electron/ghostpad-client.js
const net = require("net");

const GPAD_PORT = 6967;
const PROBE_PORTS = [6967, 9021];

let socket = null;
let connectedIp = null;
let connectedPort = null;

function buildGpadBuffer(state) {
  const buf = Buffer.alloc(16);
  buf.write("GPAD", 0);
  buf.writeUInt32BE(state.buttons >>> 0, 4);
  buf[8] = state.lx ?? 128;
  buf[9] = state.ly ?? 128;
  buf[10] = state.rx ?? 128;
  buf[11] = state.ry ?? 128;
  buf[12] = state.l2 ?? 0;
  buf[13] = state.r2 ?? 0;
  return buf;
}

function connect(ip, port = GPAD_PORT) {
  return new Promise((resolve, reject) => {
    if (socket) {
      try {
        socket.destroy();
      } catch (_) {
        /* ignore */
      }
      socket = null;
      connectedIp = null;
      connectedPort = null;
    }

    const client = net.createConnection({ host: ip, port }, () => {
      client.setNoDelay(true);
      socket = client;
      connectedIp = ip;
      connectedPort = port;
      resolve({ ip, port });
    });

    client.on("error", (err) => {
      if (socket === client) {
        socket = null;
        connectedIp = null;
        connectedPort = null;
      }
      reject(err);
    });

    client.on("close", () => {
      if (socket === client) {
        socket = null;
        connectedIp = null;
        connectedPort = null;
      }
    });
  });
}

function disconnect() {
  return new Promise((resolve) => {
    if (!socket) {
      resolve({ ok: true });
      return;
    }
    try {
      socket.write(Buffer.from("DISC"));
    } catch (_) {
      /* ignore */
    }
    socket.end();
    socket = null;
    connectedIp = null;
    connectedPort = null;
    resolve({ ok: true });
  });
}

function sendPadState(state) {
  if (!socket || socket.destroyed) {
    return { ok: false, error: "Not connected" };
  }
  try {
    socket.write(buildGpadBuffer(state));
    return { ok: true };
  } catch (err) {
    return { ok: false, error: err.message };
  }
}

function getStatus() {
  return {
    isConnected: Boolean(socket && !socket.destroyed),
    ip: connectedIp,
    port: connectedPort || GPAD_PORT,
  };
}

function probeHost(ip, port, timeoutMs = 800) {
  return new Promise((resolve) => {
    const client = new net.Socket();
    let settled = false;
    const finish = (result) => {
      if (settled) return;
      settled = true;
      try {
        client.destroy();
      } catch (_) {
        /* ignore */
      }
      resolve(result);
    };

    client.setTimeout(timeoutMs);
    client.once("connect", () => finish({ ip, port, reachable: true }));
    client.once("timeout", () => finish({ ip, port, reachable: false }));
    client.once("error", () => finish({ ip, port, reachable: false }));
    client.connect(port, ip);
  });
}

function getLocalSubnetBase() {
  const os = require("os");
  const ifaces = os.networkInterfaces();
  for (const entries of Object.values(ifaces)) {
    for (const entry of entries || []) {
      if (entry.family === "IPv4" && !entry.internal) {
        const parts = entry.address.split(".");
        if (parts.length === 4) {
          return `${parts[0]}.${parts[1]}.${parts[2]}`;
        }
      }
    }
  }
  return "192.168.1";
}

async function scanNetwork(options = {}) {
  const base = options.subnet || getLocalSubnetBase();
  const ports = options.ports || PROBE_PORTS;
  const timeoutMs = options.timeoutMs || 600;
  const found = [];
  const tasks = [];

  for (let i = 1; i <= 254; i++) {
    const ip = `${base}.${i}`;
    for (const port of ports) {
      tasks.push(
        probeHost(ip, port, timeoutMs).then((result) => {
          if (result.reachable) {
            found.push({
              ip,
              port,
              service: port === 6967 ? "ghostpad" : "elfldr",
            });
          }
        })
      );
    }
  }

  await Promise.all(tasks);

  const byIp = new Map();
  for (const hit of found) {
    const existing = byIp.get(hit.ip) || { ip: hit.ip, ports: [] };
    if (!existing.ports.includes(hit.port)) {
      existing.ports.push(hit.port);
    }
    if (hit.port === 6967) existing.hasGhostpad = true;
    if (hit.port === 9021) existing.hasElfldr = true;
    byIp.set(hit.ip, existing);
  }

  return Array.from(byIp.values()).sort((a, b) => {
    const aScore = (a.hasGhostpad ? 2 : 0) + (a.hasElfldr ? 1 : 0);
    const bScore = (b.hasGhostpad ? 2 : 0) + (b.hasElfldr ? 1 : 0);
    return bScore - aScore;
  });
}

const CTRL_PORT = 6970;

function sendCtrlPacket(ip, buf, timeoutMs = 3000) {
  return new Promise((resolve) => {
    const socket = new net.Socket();
    socket.setTimeout(timeoutMs);
    socket.once("error", (err) => resolve({ ok: false, error: err.message }));
    socket.once("timeout", () => {
      socket.destroy();
      resolve({ ok: false, error: "timeout" });
    });
    socket.connect(CTRL_PORT, ip, () => {
      socket.write(buf, () => {
        socket.end();
        resolve({ ok: true });
      });
    });
  });
}

function sendType(ip, deviceType = 3) {
  const buf = Buffer.alloc(8);
  buf.write("TYPE", 0);
  buf.writeUInt32LE(deviceType >>> 0, 4);
  return sendCtrlPacket(ip, buf);
}

function sendGpadPacket(ip, buf, timeoutMs = 3000) {
  return new Promise((resolve) => {
    const s = new net.Socket();
    s.setTimeout(timeoutMs);
    s.once("error", () => resolve({ ok: false }));
    s.once("timeout", () => { s.destroy(); resolve({ ok: false }); });
    s.connect(GPAD_PORT, ip, () => {
      s.write(buf, () => { s.end(); resolve({ ok: true }); });
    });
  });
}

function disconnectVirtual(ip) {
  const gpadDisc = Buffer.alloc(16);
  gpadDisc.write("DISC", 0);
  const ctrlDisc = Buffer.from("DISC");

  const gpadPromise = (socket && !socket.destroyed)
    ? new Promise((resolve) => {
        try {
          socket.write(gpadDisc);
          resolve({ ok: true });
        } catch {
          resolve({ ok: false });
        }
      })
    : sendGpadPacket(ip, gpadDisc);

  const ctrlPromise = sendCtrlPacket(ip, ctrlDisc).catch(() => ({ ok: false }));
  return Promise.all([gpadPromise, ctrlPromise]).then(([g, c]) => ({
    ok: g.ok || c.ok,
  }));
}

module.exports = {
  GPAD_PORT,
  CTRL_PORT,
  connect,
  disconnect,
  sendPadState,
  getStatus,
  probeHost,
  scanNetwork,
  sendType,
  disconnectVirtual,
};

File: Ghostpad-native/src/network/ghostpad_client.cpp
// Ghostpad Native - PS5 Remote Controller
// Copyright (c) 2026  seregonwar
// Based on original Ghostpad by stonedmodder  
// Licensed under the GNU General Public License v3.0. See LICENSE file for details.

#include "network/ghostpad_client.h"
#include "network/network_scanner.h"
#include "network/socket_util.h"
#include <cstring>
#include <vector>
#include <thread>
#include <algorithm>
#include <map>
#include <mutex>
#include <atomic>

namespace ghostpad {

static bool sockets_initialized = false;

static void initSockets() {
    if (sockets_initialized) return;
#ifdef _WIN32
    WSADATA wsa;
    WSAStartup(MAKEWORD(2, 2), &wsa);
#endif
    sockets_initialized = true;
}

GhostpadClient::GhostpadClient() {
    initSockets();
}

GhostpadClient::~GhostpadClient() {
    disconnect();
}

bool GhostpadClient::connect(const std::string& ip, int port, int timeout_ms) {
    /*
     *    [ NON-BLOCKING MUTEX-FREE CONNECT ]
     *    Probes connection on a local socket descriptor first to avoid locking
     *    the instance mutex during select(), which would freeze the GUI thread.
     */
    socket_t temp_sock = ::socket(AF_INET, SOCK_STREAM, 0);
    if (temp_sock == INVALID_SOCKET_VAL) return false;

    setNoSigPipe(temp_sock);
    setNonBlocking(temp_sock, true);

    struct sockaddr_in addr = {};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(static_cast<uint16_t>(port));
    inet_pton(AF_INET, ip.c_str(), &addr.sin_addr);

    int ret = ::connect(temp_sock, (struct sockaddr*)&addr, sizeof(addr));

    if (ret < 0) {
#ifdef _WIN32
        if (WSAGetLastError() != WSAEWOULDBLOCK) {
            closeSocket(temp_sock);
            return false;
        }
#else
        if (errno != EINPROGRESS) {
            closeSocket(temp_sock);
            return false;
        }
#endif
    }

    fd_set fdset;
    FD_ZERO(&fdset);
    FD_SET(temp_sock, &fdset);
    struct timeval tv;
    tv.tv_sec = timeout_ms / 1000;
    tv.tv_usec = (timeout_ms % 1000) * 1000;

    ret = select(static_cast<int>(temp_sock) + 1, nullptr, &fdset, nullptr, &tv);
    if (ret <= 0) {
        closeSocket(temp_sock);
        return false;
    }

    int so_error = 0;
    socklen_t len = sizeof(so_error);
    getsockopt(temp_sock, SOL_SOCKET, SO_ERROR, (char*)&so_error, &len);
    if (so_error != 0) {
        closeSocket(temp_sock);
        return false;
    }

    setNonBlocking(temp_sock, false);

    // Disable Nagle's algorithm
    int flag = 1;
    setsockopt(temp_sock, IPPROTO_TCP, TCP_NODELAY, (char*)&flag, sizeof(flag));

    // Lock the instance state mutex briefly only to commit the successfully opened socket
    {
        std::lock_guard<std::mutex> lock(mutex_);
        if (connected_ && sock_ != INVALID_SOCKET_VAL) {
            try {
                auto disc = buildDiscPacket();
                ::send(sock_, (const char*)disc.data(), static_cast<int>(disc.size()), 0);
            } catch (...) {}
            closeSocket(sock_);
        }
        sock_ = temp_sock;
        connected_ = true;
        host_ = ip;
        port_ = port;
    }
    return true;
}

void GhostpadClient::disconnect() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (!connected_) return;

    if (sock_ != INVALID_SOCKET_VAL) {
        try {
            auto disc = buildDiscPacket();
            ::send(sock_, (const char*)disc.data(), static_cast<int>(disc.size()), 0);
        } catch (...) {}

        closeSocket(sock_);
        sock_ = INVALID_SOCKET_VAL;
    }
    connected_ = false;
    host_.clear();
    port_ = GPAD_PORT;
}

bool GhostpadClient::sendPadState(const GpadNetworkState& state) {
    std::lock_guard<std::mutex> lock(mutex_);
    if (!connected_ || sock_ == INVALID_SOCKET_VAL) return false;

    auto packet = buildGpadPacket(state);
    int sent = ::send(sock_, (const char*)packet.data(), static_cast<int>(packet.size()), 0);
    if (sent == static_cast<int>(packet.size())) return true;

    closeSocket(sock_);
    sock_ = INVALID_SOCKET_VAL;
    connected_ = false;
    return false;
}

GhostpadStatus GhostpadClient::getStatus() const {
    std::lock_guard<std::mutex> lock(mutex_);
    GhostpadStatus s;
    s.is_connected = connected_;
    s.ip = host_;
    s.port = port_;
    return s;
}

CtrlResult GhostpadClient::sendCtrlPacket(const std::string& ip, const uint8_t* data, size_t len, int timeout_ms) {
    CtrlResult result;

    socket_t sock = ::socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET_VAL) {
        result.error = "Socket creation failed";
        return result;
    }

    setNoSigPipe(sock);
    setSocketTimeout(sock, SO_SNDTIMEO, timeout_ms);
    setSocketTimeout(sock, SO_RCVTIMEO, timeout_ms);

    struct sockaddr_in addr = {};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(static_cast<uint16_t>(CTRL_PORT));
    inet_pton(AF_INET, ip.c_str(), &addr.sin_addr);

    if (::connect(sock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        result.error = "Connection failed: " + getLastSocketErrorString();
        closeSocket(sock);
        return result;
    }

    int flag = 1;
    setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, (char*)&flag, sizeof(flag));

    int sent = ::send(sock, (const char*)data, static_cast<int>(len), 0);
    closeSocket(sock);

    if (sent == static_cast<int>(len)) {
        result.ok = true;
    } else {
        result.error = "Send failed";
    }
    return result;
}

CtrlResult GhostpadClient::sendType(const std::string& ip, uint32_t device_type, int timeout_ms) {
    auto packet = buildTypePacket(device_type);
    return sendCtrlPacket(ip, packet.data(), packet.size(), timeout_ms);
}

CtrlResult GhostpadClient::disconnectVirtual(const std::string& ip) {
    auto gpad_disc = buildDiscPacket();
    auto ctrl_disc = buildDiscPacket();

    CtrlResult r1 = sendCtrlPacket(ip, gpad_disc.data(), gpad_disc.size(), 3000);
    CtrlResult r2 = sendCtrlPacket(ip, ctrl_disc.data(), ctrl_disc.size(), 3000);

    CtrlResult result;
    result.ok = r1.ok || r2.ok;
    return result;
}

CtrlResult GhostpadClient::terminatePayload(const std::string& ip, int timeout_ms) {
    auto packet = buildUnptPacket();
    return sendCtrlPacket(ip, packet.data(), packet.size(), timeout_ms);
}

bool GhostpadClient::probeHostPort(const std::string& ip, int port, int timeout_ms) {
    socket_t sock = ::socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET_VAL) return false;

    setNonBlocking(sock, true);

    struct sockaddr_in addr = {};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(static_cast<uint16_t>(port));
    inet_pton(AF_INET, ip.c_str(), &addr.sin_addr);

    ::connect(sock, (struct sockaddr*)&addr, sizeof(addr));

    fd_set fdset;
    FD_ZERO(&fdset);
    FD_SET(sock, &fdset);
    struct timeval tv;
    tv.tv_sec = timeout_ms / 1000;
    tv.tv_usec = (timeout_ms % 1000) * 1000;

    int ret = select(static_cast<int>(sock) + 1, nullptr, &fdset, nullptr, &tv);

    bool reachable = false;
    if (ret > 0) {
        int so_error = 0;
        socklen_t len = sizeof(so_error);
        getsockopt(sock, SOL_SOCKET, SO_ERROR, (char*)&so_error, &len);
        reachable = (so_error == 0);
    }

    closeSocket(sock);
    return reachable;
}

std::vector<ScanResult> GhostpadClient::scanNetwork(const std::string& subnet, int timeout_ms) {
    std::string base = subnet.empty() ? getLocalSubnet() : subnet;
    std::vector<ScanResult> scanned;

    /*
     *    [ MULTI-THREADED SUBNET SCANNER ]
     *    Uses worker threads to check IPs concurrently.
     */
    std::vector<ProbeResult> hits;
    std::mutex hits_mutex;
    std::atomic<int> next_ip{1};

    const int NUM_THREADS = 64;
    std::vector<std::thread> workers;
    for (int t = 0; t < NUM_THREADS; t++) {
        workers.emplace_back([&]() {
            while (true) {
                int i = next_ip++;
                if (i > 254) break;
                
                std::string host = base + "." + std::to_string(i);
                for (int port : PROBE_PORTS) {
                    if (probeHostPort(host, port, timeout_ms)) {
                        std::lock_guard<std::mutex> lock(hits_mutex);
                        hits.push_back({host, port, true});
                    }
                }
            }
        });
    }

    for (auto& w : workers) {
        w.join();
    }

    // Group by IP
    std::map<std::string, ScanResult> by_ip;
    for (auto& hit : hits) {
        auto& entry = by_ip[hit.ip];
        entry.ip = hit.ip;
        bool found = false;
        for (auto p : entry.ports) {
            if (p == hit.port) { found = true; break; }
        }
        if (!found) entry.ports.push_back(hit.port);
        if (hit.port == 6967) entry.has_ghostpad = true;
        if (hit.port == 9021 || hit.port == 9090) entry.has_elfldr = true;
    }

    for (auto& [_, v] : by_ip) {
        scanned.push_back(v);
    }

    std::sort(scanned.begin(), scanned.end(), [](const ScanResult& a, const ScanResult& b) {
        int sa = (a.has_ghostpad ? 2 : 0) + (a.has_elfldr ? 1 : 0);
        int sb = (b.has_ghostpad ? 2 : 0) + (b.has_elfldr ? 1 : 0);
        return sa > sb;
    });

    return scanned;
}

std::string GhostpadClient::getLocalSubnet() {
    return NetworkScanner::getLocalSubnet();
}

} // namespace ghostpad
